### ISSUE TITLE
Add readinessProbe and livenessProbe support for target allocator 

### ISSUE THREAD

**Issue by luckyxiaoqiang** [open-telemetry/opentelemetry-operator#2258]
### Component(s)

target allocator

### Is your feature request related to a problem? Please describe.

To acheive HA, we needs readinessProbe and livenessProbe configuration. 

### Describe the solution you'd like

Add readinessProbe and livenessProbe properties to `spec.targetAllocator`, and pass it to target allocator `Deployment`.

**Comment by changexd** [comment]
@jaronoff97 Hi I'm interested in this issue, I would like to help this out if this is still available, after some testing and a bit of research, I have couple of question for specification I would like to know before making changes that may not suit the need.

Target Allocator:
1. Should liveness and readiness be checked separately or simply a HTTPGET to "/health" endpoint for both, what should be returned when health checked(currently I'm only writing 200 to the header)
2. Is there any specific condition should be met before checking or simply an endpoint checking the server is running or not?

Operator:
1. Is additional field needed for target allocator spec for customizing probe parameters? if so, what parameters should be inside the field, if not, is it okay to simply add a default probe parameters to target allocator deployment when generating.

If there's already an idea/direction how this should be implemented, please feel free to let me know, thank you : )

**Comment by jaronoff97** [comment]
@changexd for your questions:

* liveness and readiness should be separate IMO
  * liveness can be a very simple check that an endpoint just returns a 200
  * readiness can be defined as after the TA has finished initializing its configuration IMO (cc @swiatekm-sumo who may have more thoughts here)
  * this [blog post](https://developers.redhat.com/blog/2020/11/10/you-probably-need-liveness-and-readiness-probes) will have some more info 
* Hm... I think for the probe we can set a default for now (this should be done in a follow up PR to implementing the endpoints)
  * I want to avoid adding these in to be customizable for now solely because we are in the midst of working on our v2 types for these resources and this is going to add more fields to the target allocator, which I think should be avoided for now.

**Comment by changexd** [comment]
Sorry, I think I mistakenly created separate PRs for readiness and liveness, should I create another PR for readiness or just edit these PRs?

**Comment by swiatekm-sumo** [comment]
As a rule, smaller PRs are better, so I'm fine with what you've done.

**Comment by rashmichandrashekar** [comment]
> Agreed, we should add the probes, but only make them configurable in the separate TargetAllocator CRD once that lands.
> 
> For the TA readiness probe, if we want to actually wait until some scrape configs are available, we'd need to add that to the HTTP server. I think just checking whether `Server.scrapeConfigResponse` is nil would be good enough, but we'd have to verify if this works correctly in the case where there aren't any scrape configs.

@swiatekm-sumo - Is there any draft of the TA CRD that is being planned? Was just curious how we could configure the liveness probe for CRD in the future.

**Comment by jaronoff97** [comment]
@rashmichandrashekar not yet, though i can tell you it will be the same as the configuration for the collector currently.

**Comment by rashmichandrashekar** [comment]
Thanks @jaronoff97. So I were to have my own go exe in the TA container, I can just add it to the liveness and the livez endpoint will honor it?

**Comment by jaronoff97** [comment]
I'm not sure what you mean, but essentially we will eventually expose the ability to configure the liveness and readiness probes like you would other kube pods
